### TITANIC ANALYSIS

#### QUESTION

1. Giới tính có ảnh hưởng quyết định đến việc sống sót không? 
    - Dự đoán: Nữ có tỉ lệ sống sót cao hơn nam.

2. Câu hỏi: Hạng vé có sự phân hóa rõ rệt về tỉ lệ sống sót không? 
    - Dự đoán: Hạng 1 có tỉ lệ sống sót cao nhất.

3. Câu hỏi: Số lượng anh chị em/vợ chồng đi cùng có giúp hành khách dễ sống sót hơn việc đi một mình không? 
    - Dự đoán: Những người đi cùng có tỉ lệ sống sót cao hơn người đi một mình.

4. Câu hỏi: Những người đi cùng cha mẹ hoặc con cái có tỉ lệ cứu sống cao hơn không? 
    - Dự đoán: Có, các gia đình thường cố gắng bám sát nhau và được ưu tiên lên thuyền cứu sinh.

5. Câu hỏi: Tổng số người thân trên tàu quá đông có làm giảm khả năng thoát hiểm không?
    - Dự đoán: Có, gia đình quá đông người sẽ khó tập hợp đầy đủ nên tỉ lệ sống sót sẽ giảm.

6. Câu hỏi: Giá vé trung bình của những người sống sót có cao hơn những người không sống sót không?
    - Dự đoán: Có, những người sống sót thường là những người đã trả giá vé (`Fare`) cao hơn (thường ở hạng 1).

7. Câu hỏi: Việc thông tin số phòng bị thiếu rất nhiều có liên quan đến việc họ không sống sót để khai báo lại không?
    - Dự đoán: Những hành khách có thông tin Cabin rõ ràng sẽ có tỉ lệ sống sót cao hơn nhóm bị trống dữ liệu Cabin.

8. Câu hỏi: Có sự tương quan nào giữa độ tuổi và hạng vé không?
    - Dự đoán: Hành khách hạng 1 sẽ có độ tuổi trung bình cao hơn hẳn hành khách hạng 3.

9. Câu hỏi: Hành khách lớn tuổi nhất trên tàu có sống sót không?
    - Dự đoán: Hành khách lớn tuổi nhất có khả năng cao là không sống sót do sức khỏe yếu.

10. Câu hỏi: Những người có cùng số vé - tức là đi chung một nhóm - có cùng chung số phận (cùng sống hoặc cùng chết) không?
    - Dự đoán: Những người chung mã vé thường có kết cục giống nhau.

#### OVERVIEW

In [ ]:
import seaborn as sns

# Load dataset
titanic = sns.load_dataset('titanic')

titanic.head()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


Nhìn vào 5 hành khách đầu tiên, ta thấy ngay hành khách số 0 là nam, 22 tuổi, đi hạng 3 và không sống sót (survived = 0), trong khi hành khách số 1 là nữ, đi hạng 1 và đã sống sót. Dữ liệu cũng cho thấy cột cabin bắt đầu xuất hiện các giá trị NaN, báo hiệu việc thiếu hụt dữ liệu ở thuộc tính này.

In [3]:
titanic.tail()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.00,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.00,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.45,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.00,C148,C
890,891,0,3,"Dooley, Mr. Patrick",male,32.0,0,0,370376,7.75,NaN,Q


Các dòng cuối cùng xác nhận tập dữ liệu kết thúc ở chỉ số (index) 890, khớp hoàn toàn với tổng số 891 hành khách đã biết. Dữ liệu phần cuối cũng cho thấy một xu hướng phổ biến: có rất nhiều nam giới đi khoang hạng 3 (Third class), đi một mình (alone = True) và phần lớn trong số họ có kết cục là không sống sót.

In [4]:
titanic.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 83.7 KB


Kết quả từ info() cho thấy dữ liệu đang bị khuyết ở một số cột quan trọng: cột age bị thiếu khoảng 177 giá trị, và đặc biệt cột deck bị thiếu dữ liệu rất trầm trọng. Ngoài ra, các biến số đã được phân loại rất rõ ràng thành các kiểu dữ liệu phù hợp như float64 (giá vé, tuổi), int64 và category để tối ưu bộ nhớ.

In [5]:
titanic.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


Nhìn vào giá trị trung bình (mean) của cột survived, ta có thể thấy chỉ có khoảng 38.3% trong số 891 hành khách này sống sót. Về độ tuổi, độ tuổi trung bình trên tàu là gần 30 tuổi, hành khách nhỏ tuổi nhất là một em bé khoảng 5 tháng tuổi (0.42), lớn tuổi nhất là 80 tuổi, và có người đã chi tới hơn 512 đô la để mua vé lên tàu.

In [6]:
titanic.shape

(891, 12)

Tập dữ liệu có tổng cộng 891 hàng, đại diện cho 891 hành khách có mặt trên chuyến tàu Titanic. Mỗi hành khách được mô tả bởi 15 cột thông tin khác nhau.

#### MISSING VALUE

In [ ]:
import pandas as pd

# Load dataset
file_path = '../../Data/Slot_01/train.csv'
titanic = pd.read_csv(file_path)

# TÍnh số lượng và phần trăm dữ liệu bị thiếu 
missing_count = titanic.isnull().sum()
missing_percent = (missing_count / len(titanic)) * 100

# Tạo DataFrame để hiển thị
missing_data = pd.DataFrame(
    {
        'Số lượng bị thiếu' : missing_count,
        'Phần trăm (%)' : missing_percent
    }
)

# Chỉ lọc ra các cột dữ liệu bị thiếu và sắp xếp từ cao xuống
missing_data = missing_data[missing_data['Số lượng bị thiếu'] > 0].sort_values(by="Phần trăm (%)", ascending=False)

missing_data

,Số lượng bị thiếu,Phần trăm (%)
Cabin,687,77.104377
Age,177,19.865320
Embarked,2,0.224467


Assess: 
- Cột cabin thiếu trên 77% dữ liệu.
- Cột age thiếu trên 19% dữ liệu.
- Cột embarked chỉ thiếu 2 hành khách. Có thể dễ dàng xử lý.

#### CATEGORY ANALYSIS

In [ ]:
import pandas as pd

file_path = '../../Data/Slot_01/train.csv'
titanic = pd.read_csv(file_path)

# Đếm số lượng hành khách theo giới tính
print("--- Phân bố Giới tính (Sex) ---")
print(titanic['Sex'].value_counts())

# Đếm số lượng hành khách theo hạng vé
print("\n--- Phân bố Hạng vé (Pclass) ---")
print(titanic['Pclass'].value_counts())

# Đếm số lượng hành khách theo cảng khởi hành (Bao gồm cả giá trị bị thiếu)
print("\n--- Phân bố Cảng khởi hành (Embarked) ---")
print(titanic['Embarked'].value_counts(dropna=False))

--- Phân bố Giới tính (Sex) ---
Sex
male      577
female    314
Name: count, dtype: int64

--- Phân bố Hạng vé (Pclass) ---
Pclass
3    491
1    216
2    184
Name: count, dtype: int64

--- Phân bố Cảng khởi hành (Embarked) ---
Embarked
S      644
C      168
Q       77
NaN      2
Name: count, dtype: int64


Assess:
- Nam giới chiếm tỷ lệ áo đảo phụ nữ
- Số lượng hàng khách của hạng vé 3 chiếm nhiều hơn so với hạng vé 1 và 2
- Cảng S  là cảng xuất phát chính khi có số lượng hàng khách nhiều hơn so với cảng Q và cảng C

#### NUMERICS ANALYSIS

In [ ]:
import pandas as pd

file_path = '../../Data/Slot_01/train.csv'
titanic = pd.read_csv(file_path)

numericals_cols = ['Age', 'Fare', 'SibSp']

stats = titanic[numericals_cols].agg(['mean', 'median', 'min', 'max'])
print(stats)

              Age        Fare     SibSp
mean    29.699118   32.204208  0.523008
median  28.000000   14.454200  0.000000
min      0.420000    0.000000  0.000000
max     80.000000  512.329200  8.000000


Assess:
- Age: Không có điểm quá bất thường về mặt logic, độ tuổi trung bình khoảng 29 tuổi, có trẻ sơ sinh và người lớn tuổi nhất là 80 tuổi.
- Fare: Giá trung bình là 32$, lớn hơn gấp đôi so với median. Nguyên nhân là do một số vé cực đắt (lên tới 512$) trong khi vẫn có các vé free (0$). Đây là một điểm bất thường, có thể cần được xử lý trong quá trình tiếp theo
- SibSp: Số liệu chỉ ra rằng, hầu hết hành khách đi một mình. Tuy nhiên max là 8, là 1 gia đình rất đông người là 1 điều hiếm gặp.

#### SURVIVAL RATE ANALYSIS

In [20]:
import seaborn as sns
titanic = sns.load_dataset('titanic')

# Tính tỉ lệ sống sót toàn tàu
survival_rate = titanic['survived'].mean()
print(f"Tỉ lệ sống sót toàn tàu: {survival_rate:.2%}")

# Xem số lượng chi tiết
print("\nSố lượng hành khách sống sót (1) và tử nạn (0):")
print(titanic['survived'].value_counts())

Tỉ lệ sống sót toàn tàu: 38.38%

Số lượng hành khách sống sót (1) và tử nạn (0):
survived
0    549
1    342
Name: count, dtype: int64


Assess:
- Tỉ lệ sống sót của toàn bộ hành khách trên tàu Titanic chỉ đạt 38.38%. 
- Con số này đồng nghĩa với việc có tới hơn 549 người đã tử nạn trong vụ đắm tàu kinh điển này.
- Thống kê cho thấy sự khốc liệt của thảm họa Titanic, phản ánh sự thiếu hụt nghiêm trọng số lượng thuyền cứu sinh và công tác cứu hộ đã gặp nhiều khó khăn.

#### ANALYSIS BY SEX & TICKET CLASS

In [24]:
import seaborn as sns
titanic = sns.load_dataset('titanic')

# Tỉ lệ sống sót
survival_combined = titanic.groupby(['sex', 'pclass'])['survived'].mean()

print("--- Tỉ lệ sống sót ---")
print(survival_combined.unstack() * 100)

--- Tỉ lệ sống sót ---
pclass          1          2          3
sex                                    
female  96.808511  92.105263  50.000000
male    36.885246  15.740741  13.544669


Asssess:
- Tỉ lệ sống sót của phụ nữ có vé hạn 1 cao nhất gần 97%.
- Tỉ lệ sống sót của nam giới ở tất cả các hạng vé đều rất thấp, cho thấy rằng giới tính là một yếu tố quyết định đến khả năng sống sót của hành khách trên tàu Titanic.

#### 3 QUESTIONS 

In [ ]:
# Câu 3: Số lượng người đi cùng (SibSp) vs Sống sót
print("--- Câu 3: Tỉ lệ sống sót theo việc có đi cùng người thân ---")
# True nếu sibsp > 0 (có đi cùng), False nếu đi một mình
is_accompanied = titanic['sibsp'] > 0
print(titanic.groupby(is_accompanied)['survived'].mean())

# Câu 6: Giá vé trung bình vs Sống sót
print("\n--- Câu 6: Giá vé trung bình theo tình trạng sống sót ---")
print(titanic.groupby('survived')['fare'].mean())

# Câu 9: Hành khách lớn tuổi nhất có sống sót không?
print("\n--- Câu 9: Tình trạng của hành khách lớn tuổi nhất ---")
# Lọc ra dòng dữ liệu của người có số tuổi bằng với số tuổi max
oldest_passenger = titanic[titanic['age'] == titanic['age'].max()]
print(oldest_passenger[['age', 'sex', 'pclass', 'survived']])

--- Câu 3: Tỉ lệ sống sót theo việc có đi cùng người thân ---
sibsp
False    0.345395
True     0.466431
Name: survived, dtype: float64

--- Câu 6: Giá vé trung bình theo tình trạng sống sót ---
survived
0    22.117887
1    48.395408
Name: fare, dtype: float64

--- Câu 9: Tình trạng của hành khách lớn tuổi nhất ---
      age   sex  pclass  survived
630  80.0  male       1         1


Assess:
- Câu 3 [ĐÚNG]: Dữ liệu cho thấy những người đi một mình (`False`) có tỉ lệ sống sót chỉ khoảng 34.5%. Trong khi đó, những người đi cùng ít nhất 1 người thân (`True`) có tỉ lệ sống sót cao hơn đáng kể, đạt khoảng 46.6%.

- Câu 6 [ĐÚNG]:  Giá vé trung bình của nhóm người sống sót khoảng 48.4 đô la, cao hơn gấp đôi so với mức giá trung bình của nhóm tử nạn là khoảng 22.1 đô la. 

- Câu 9 [SAI]:  Hành khách lớn tuổi nhất trên tàu là một người đàn ông 80 tuổi, đi khoang hạng 1. Dữ liệu ghi nhận ông ấy đã survived.


#### NEW QUESTIONS

1. C1 - Liệu tỉ lệ sống sót của các nhóm tuổi (trẻ em, người trưởng thành, người già) có xu hướng tăng tiến rõ rệt khi chuyển dần từ hạng vé thấp (Hạng 3) lên hạng vé cao (Hạng 1) không?
2. C2 - Qua thống kê, có một đại gia đình đi cùng nhau tới 8 người (`SibSp = 8`). Số phận của gia đình đông đúc nhất chuyến tàu này ra sao, có ai sống sót không?
3. C3 - Những hành khách mua chiếc vé đắt đỏ nhất trên tàu (lên tới 512.33 đô la) có tỉ lệ sống sót là 100%, hay vẫn có người tử nạn?
4. C4 - Nhóm hành khách có giá vé bằng 0 (`Fare = 0`) thực chất là ai (đàn ông hay phụ nữ, độ tuổi nào), và họ có được ưu tiên cứu hộ không?
5. C5 - Những hành khách bị thiếu thông tin độ tuổi có tỉ lệ sống sót là bao nhiêu?

#### C1 - Combined survival rate by Age Group and Ticket Class

In [33]:
titanic_temp = titanic.copy()

# Chia nhóm tuổi: Trẻ em (0-12), Trưởng thành (13-59), Người già (60-100)
age_bins = [0, 12, 59, 100]
age_labels = ['Trẻ em', 'Trưởng thành', 'Người già']
titanic_temp['age_group'] = pd.cut(titanic_temp['age'], bins=age_bins, labels=age_labels)

# Groupby theo Hạng vé và Nhóm tuổi
survival_temp = titanic_temp.groupby(['pclass', 'age_group'], observed=False)['survived'].mean().unstack()

print("--- TỈ LỆ SỐNG SÓT THEO HẠNG VÉ VÀ NHÓM TUỔI ---")
print(survival_temp*100)


--- TỈ LỆ SỐNG SÓT THEO HẠNG VÉ VÀ NHÓM TUỔI ---
age_group      Trẻ em  Trưởng thành  Người già
pclass                                        
1           75.000000     69.090909  29.411765
2          100.000000     42.763158  25.000000
3           41.666667     21.192053  20.000000


Assess:
- Xu hướng tăng rõ rệt: Đúng như dự đoán, tỉ lệ sống sót của mọi nhóm tuổi đều tăng mạnh khi chuyển từ hạng vé thấp (Hạng 3) lên hạng vé cao (Hạng 1).
- Note: Đặc quyền của trẻ em - Trẻ em ở khoang Hạng 1 và Hạng 2 được bảo vệ gần như tuyệt đối (tỉ lệ sống sót 100% ở hạng 2 và ~75.7% ở hạng 1). Tuy nhiên, trẻ em ở khoang Hạng 3 lại chỉ có khoảng ~41% sống sót. 
- Hạng vé có thể là yếu tố quyết định cơ hội sống của mọi lứa tuổi.

In [48]:
# Lọc danh sách khách hàng có giá vé bằng 0
zero_fare = titanic[titanic['fare'] == 0]

# Tính toán các chỉ số
total_passengers = len(zero_fare)
gender_counts = zero_fare['sex'].value_counts().to_string(header=False)
min_age = zero_fare['age'].min()
max_age = zero_fare['age'].max()
survival_rate = zero_fare['survived'].mean()

# Hiển thị kết quả
print("--- THỐNG KÊ KHÁCH HÀNG VÉ 0 ĐỒNG ---")
print(f"- Tổng số lượng : {total_passengers} người")
print(f"- Giới tính     : {gender_counts.replace(chr(10), chr(10))}")
print(f"- Độ tuổi       : {min_age:.1f} - {max_age:.1f} tuổi")
print(f"- Tỉ lệ sống sót: {survival_rate:.2%}")

--- THỐNG KÊ KHÁCH HÀNG VÉ 0 ĐỒNG ---
- Tổng số lượng : 15 người
- Giới tính     : male    15
- Độ tuổi       : 19.0 - 49.0 tuổi
- Tỉ lệ sống sót: 6.67%


Assess:
- Họ là: Dữ liệu cho thấy có tổng cộng 15 người có vé 0$ trên tập dữ liệu này. Và 100% trong số họ là nam giới trong độ tuổi lao động (từ 19 đến 49 tuổi). Có thể là các thủy thủ đoàn, nhân viên bảo hành tàu hoặc những người được cấp vé danh dự
- Họ có được ưu tiên không? KHÔNG. Tỉ lệ sống sót của nhóm này cực kỳ thê thảm, chỉ đạt 6.67% (chỉ có 1/15 người sống sót).